src/dataset.js

```diff
@@ -33,16 +33,36 @@ class State {
   }
 
   get(index) {
+    if(index > this.length) {return null;}
     let pageOffset = Math.floor(index / this.pageSize);
     let recordOffset = index % this.pageSize;
+
+    // Dynamically find the page offset
+    const minUnloadPage = Math.floor((this.readOffset - this.unloadHorizon) / this.pageSize);
+    const minUnloadHorizon = Math.max(minUnloadPage, 0);
+    // const maxUnloadPage = Math.ceil((this.readOffset  + this.unloadHorizon) / this.pageSize);
+    // const maxUnloadHorizon = Math.min(this.stats.totalPages || Infinity, maxUnloadPage, this.pages.length);
+    if(pageOffset >= minUnloadHorizon) {
+      let _this = {index: index};
+      pageOffset = this.pages.findIndex(function(page) {
```

Comment:
We should be able to execute this find inside the slice of pages between `minUnloadPage` and `minUnloadHorizon`